tool_path_planner/src/edge_path_generator.cpp

```diff
@@ -0,0 +1,589 @@
+    rot.block(2,0,1,3) = Vector3d(vx.z(), vy.z(), vz.z()).array().transpose();
+    return rot;
+  };
+
+  std::vector<int> cloud_indices;
+  if(indices.empty())
+  {
+    cloud_indices.resize(cloud_normals.size());
+    std::iota(cloud_indices.begin(), cloud_indices.end(), 0);
+  }
+  else
+  {
+    cloud_indices.assign(indices.begin(), indices.end());
+  }
+
+  for(std::size_t i = 0; i < cloud_indices.size() - 1; i++)
+  {
+    std::size_t idx_current = cloud_indices[i];
+    std::size_t idx_next = cloud_indices[i+1];
+    if(idx_current >= cloud_normals.size() || idx_next >= cloud_normals.size())
+    {
+      CONSOLE_BRIDGE_logError("Invalid indices (current: %lu, next: %lu) for point cloud were passed",
+                              idx_current, idx_next);
+      return false;
+    }
+    const PointNormal& p1 = cloud_normals[idx_current];
+    const PointNormal& p2 = cloud_normals[idx_next];
+    x_dir = (point_to_vec(p2) - point_to_vec(p1)).normalized().cast<double>();
+    z_dir = Vector3d(p1.normal_x, p1.normal_y, p1.normal_z).normalized();
+    y_dir = z_dir.cross(x_dir).normalized();
+
+    pose = Translation3d(point_to_vec(p1));
+    pose.matrix().block<3,3>(0,0) = to_rotation_mat(x_dir, y_dir, z_dir);
+    tf::poseEigenToMsg(pose,pose_msg);
+
+    poses.poses.push_back(pose_msg);
+  }
+
+  // last pose
+  pose_msg = poses.poses.back();
+  pose_msg.position.x = cloud_normals[cloud_indices.back()].x;
+  pose_msg.position.y = cloud_normals[cloud_indices.back()].y;
+  pose_msg.position.z = cloud_normals[cloud_indices.back()].z;
+  poses.poses.push_back(pose_msg);
+
+  return true;
+}
+
+namespace tool_path_planner
+{
+EdgePathGenerator::EdgePathGenerator()
+{
+
+}
+
+EdgePathGenerator::~EdgePathGenerator()
+{
+
+}
+
+boost::optional< std::vector<geometry_msgs::PoseArray >> EdgePathGenerator::generate(const tool_path_planner::EdgePathConfig& config)
+{
+  using namespace pcl;
+
+  if(!input_points_ || !input_cloud_ || !mesh_)
+  {
+    CONSOLE_BRIDGE_logError("No input mesh data has been set");
+    return boost::none;
+  }
+
+  // initializing octree
+  octree_search_ = boost::make_shared< octree::OctreePointCloudSearch<PointXYZ> >(config.octree_res);
+  octree_search_->setInputCloud (input_points_);
+  octree_search_->addPointsFromInputCloud ();
+
+  // initializing edge points
+  PointCloud<PointXYZHSV>::Ptr edge_results = boost::make_shared<PointCloud<PointXYZHSV>>();
+  pcl::copyPointCloud(*input_points_, *edge_results);
+
+  // identify high v values
+  const int num_threads = config.num_threads;
+  long cnt = 0;
+  long percentage = 0;
+  long num_points = static_cast<long>(input_points_->size());
+  CONSOLE_BRIDGE_logInform("Computing Eigen Values");
+#pragma omp parallel for num_threads(num_threads)
+  // TODO avoid doing this for every points and used the returned indices 'point_idx' to
+  //      remove the points that have already been visited
+  for (std::size_t idx = 0; idx < input_points_->points.size(); ++idx)
```

Comment:
I recently ran into this but int should be used instead of size_t when using omp.

Reply:
I'll change it but it be nice if you could provide a link to this info

Reply:
We ran into when trying to build descartes_light on windows and it was getting compiler error. Here is a [link](https://stackoverflow.com/questions/3988470/why-must-loop-variables-be-signed-in-a-parallel-for) to a discussion on the topic with links to the omp specification.